lib/AST/Module.cpp

```diff
@@ -2566,6 +2566,20 @@ void
 SourceFile::setImports(ArrayRef<AttributedImport<ImportedModule>> imports) {
   assert(!Imports && "Already computed imports");
   Imports = getASTContext().AllocateCopy(imports);
+
+  // Find and cache the import of the underlying module, if present.
+  auto parentModuleName = getParentModule()->getName();
+  for (auto import : imports) {
+    if (!import.options.contains(ImportFlags::Exported))
+      continue;
+
+    auto importedModule = import.module.importedModule;
+    if (importedModule->getName() == parentModuleName &&
+        importedModule->findUnderlyingClangModule()) {
+      ImportedUnderlyingModule = import.module.importedModule;
+      break;
+    }
+  }
```

Comment:
Rather than looping over the imports this way, you could modify `ImportResolver::getModule()` to save off the underlying clang module in a new member variable of `ImportResolver` (there's a separate code path to look it up—look for the use of `getClangModuleLoader()`) and then have `performImportResolution()` set that on the `SourceFile` immediately before/after calling `setImports()`. Would that be a cleaner solution?

Reply:
That does seem like a cleaner solution because it keeps the logic for identifying the underlying clang module more centralized. In the interest of time I think I'm going to land this solution first, but I will follow it up with what you've suggested - thanks!

Reply:
Implemented here: https://github.com/apple/swift/pull/71330